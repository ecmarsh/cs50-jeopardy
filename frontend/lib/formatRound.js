export default function(page) {
  let title;
  switch (page) {
    case 1:
      title = 'Single';
      break;
    case 2:
      title = 'Double';
      break;
    case 3:
      title = 'Final';
      break;
    default:
      title = 'CS50';
  }

  return title + ' Jeopardy';
}
