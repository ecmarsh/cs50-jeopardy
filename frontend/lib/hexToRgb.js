export default function(hex) {
  const chunk = 2,
    radix = 16,
    rgb = [];

  let i = 1;
  while (hex && i < hex.length) {
    rgb.push(parseInt(hex.substring(i, i + chunk), radix));
    i += chunk;
  }

  return rgb.toString();
}
