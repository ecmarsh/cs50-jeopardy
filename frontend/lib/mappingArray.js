export default function(N = 0, isFilled = true, startAtZero = false) {
  if (N == 0) return [];
  if (N > 0 && !isFilled) return Array(N).fill();
  if (isFilled && !startAtZero)
    return Array.from({ length: N }, (v, k) => k + 1);
  return Array.from({ length: N }, (v, k) => k++);
}
