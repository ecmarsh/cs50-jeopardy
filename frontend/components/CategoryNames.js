import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import * as R from 'ramda';
import formatRound from '../lib/formatRound';
import Error from './ErrorMessage';
import Form from './styles/Form';
import Button from './styles/Button';
import { TextMuted, Divider } from './styles/Utilities';

const UPDATE_CATEGORY_MUTATION = gql`
  mutation UPDATE_CATEGORY_MUTATION($categoryId: ID!, $categoryName: String) {
    updateCategory(id: $categoryId, name: $categoryName) {
      id
      name
    }
  }
`;

class CategoryNames extends Component {
  static propTypes = {
    categories: PropTypes.array.isRequired,
  };

  state = {
    categories: [],
    called: false,
  };

  componentDidMount() {
    const { categories } = this.props;
    this.setState({ categories });
  }

  handleChange = e => {
    e.preventDefault();
    const { id, value } = e.target;
    const categories = [...this.state.categories];
    const idx = R.findIndex(R.propEq('id', id))(categories);
    categories[idx].name = value;
    categories[idx].updated = true; // flag to include in mutation
    this.setState({ categories });
  };

  handleSubmit = (e, updateCategoryMutation) => {
    // Stop default form submission
    e.preventDefault();
    // Get list of changed categories to update
    const updatedCategories = R.filter(R.has('updated'), this.state.categories);
    // Update category mutation
    const doMutation = async category => {
      const { id, name } = category;
      const res = await updateCategoryMutation({
        variables: { categoryId: id, categoryName: name },
      });
    };
    // Invoke mutation on each updated category
    R.forEach(doMutation, updatedCategories);
    // Called to flash saved message
    this.setState({ called: true });

    setTimeout(this.flashUpdatedConfirmation, 2000);
  };

  flashUpdatedConfirmation = () => {
    this.setState({ called: false });
  };

  render() {
    const { categories, called } = this.state;
    return (
      <Mutation mutation={UPDATE_CATEGORY_MUTATION}>
        {(updateCategory, { loading, error }) => {
          return (
            <Form
              data-test="category-names-form"
              className="noBg"
              onSubmit={e => this.handleSubmit(e, updateCategory)}
            >
              <h2>Category Names</h2>
              <Error error={error} />
              <fieldset disabled={loading} aria-busy={loading}>
                {categories.map((category, i) => (
                  <div key={category.id}>
                    {i === 0 && <h4>{formatRound(1)}</h4>}
                    {i === 5 && <h4>{formatRound(2)}</h4>}
                    {i === 10 && <h4>{formatRound(3)}</h4>}
                    <label
                      htmlFor={`category-${i}`}
                      style={{
                        flexDirection: 'row',
                        width: '100%',
                      }}
                    >
                      <TextMuted
                        style={{ textAlign: 'center', flexBasis: '25%' }}
                      >
                        CATEGORY {i + 1}:
                      </TextMuted>
                      <input
                        type="text"
                        name={`category-${i}`}
                        id={category.id}
                        value={category.name}
                        onChange={e => this.handleChange(e)}
                        style={{
                          flexBasis: '60%',
                          width: 'auto',
                        }}
                      />
                    </label>
                  </div>
                ))}
                <div style={{ margin: '0 auto' }}>
                  <TextMuted
                    data-test="category-names-save-confirmation"
                    hidden={!called}
                  >
                    Category Names Saved!
                  </TextMuted>
                  <Button
                    primary
                    type="submit"
                    style={{ margin: '1rem auto 2rem' }}
                  >
                    Save Category Names
                  </Button>
                  <Divider />
                  {this.props.children}
                </div>
              </fieldset>
            </Form>
          );
        }}
      </Mutation>
    );
  }
}

export default CategoryNames;
export { UPDATE_CATEGORY_MUTATION };
