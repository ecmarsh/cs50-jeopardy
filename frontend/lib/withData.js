import withApollo from 'next-with-apollo';
import ApolloClient from 'apollo-boost';
import { endpoint, endpointProd } from '../config';
import { LOCAL_STATE_QUERY } from '../components/SoundEffect';
import { ANSWERED_QUESTIONS_QUERY } from '../components/GameScreen';
import gql from 'graphql-tag';

// Setup for SSR w/ next and config endpoint
function createClient({ headers }) {
  const typeDefs = gql`
    type StudyQuestion {
      data: Json
    }
  `;
  const client = new ApolloClient({
    uri: process.env.NODE_ENV === 'development' ? endpoint : endpointProd,
    request: operation => {
      operation.setContext({
        fetchOptions: {
          credentials: 'include',
        },
        headers,
      });
    },
    typeDefs,
    // cache
    clientState: {
      resolvers: {
        Mutation: {
          toggleSound(_, variables, { cache }) {
            // read the soundOn value from the cache
            const { soundOn } = cache.readQuery({
              query: LOCAL_STATE_QUERY,
            });
            // Toggle sound on || off
            const data = {
              data: { soundOn: !soundOn },
            };
            cache.writeData(data);
            return data;
          },
          addAnswered(_, { id, val }, { cache }) {
            // read the answered questions from the cache
            const { answeredQuestions } = cache.readQuery({
              query: ANSWERED_QUESTIONS_QUERY,
            });
            const newQuestion = { id, val, __typename: 'StudyQuestion' };
            const jsonified = JSON.stringify([
              newQuestion,
              ...answeredQuestions,
            ]);
            // Add new question to data
            const data = {
              data: {
                answeredQuestions: [newQuestion, ...answeredQuestions],
              },
            };

            cache.writeData(data);
            return data;
          },
          resetBoard(_, variables, { cache }) {
            // Reset answered questions
            const data = {
              data: {
                answeredQuestions: [],
              },
            };
            cache.writeData(data);
            return data;
          },
        },
      },
      defaults: {
        soundOn: false,
        answeredQuestions: [],
      },
    },
  });
  return client;
}

export default withApollo(createClient);
