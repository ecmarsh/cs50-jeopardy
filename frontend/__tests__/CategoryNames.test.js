import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import wait from 'waait';
import { MockedProvider } from 'react-apollo/test-utils';
import CategoryNames, {
  UPDATE_CATEGORY_MUTATION,
} from '../components/CategoryNames';
import { fakeGame } from '../lib/testUtils';

describe('<CategoryNames />', () => {
  it('renders edit categories form and matches snapshot', async () => {
    const { categories } = fakeGame();
    const wrapper = mount(
      <MockedProvider>
        <CategoryNames categories={categories} />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    const form = wrapper.find('form[data-test="category-names-form"]');
    expect(toJSON(form)).toMatchSnapshot();
  });

  it('adds changed inputs to state', async function() {
    const { categories } = fakeGame();
    const wrapper = mount(
      <MockedProvider>
        <CategoryNames categories={categories} />
      </MockedProvider>
    );
    wrapper.find('input#cat-1').simulate('change', {
      target: { id: 'cat-1', value: 'An Updated Category Name' },
    });
    wrapper.update();
    const updatedCategories = JSON.parse(JSON.stringify(categories)); // deep clone
    updatedCategories[0].name = 'An Updated Category Name';
    expect(wrapper.find('CategoryNames').instance().state).toMatchObject({
      categories: updatedCategories,
      called: false,
    });
  });

  it('shows saved category names confirmation on submit', async () => {
    const { categories } = fakeGame();
    const mocks = [
      {
        request: {
          query: UPDATE_CATEGORY_MUTATION,
          variables: {
            categoryId: 'cat-1',
            categoryName: 'An Updated Category Name',
          },
        },
        result: {
          data: {
            updateCategory: {
              id: 'cat-1',
              name: 'An Updated Category Name',
              __typename: 'Category',
            },
          },
        },
      },
    ];

    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <CategoryNames categories={categories} />
      </MockedProvider>
    );

    const form = wrapper.find('form[data-test="category-names-form"]');
    form.simulate('submit');

    await wait(0);
    wrapper.update();

    const uiConfirm = wrapper.find(
      'span[data-test="category-names-save-confirmation"]'
    );
    expect(uiConfirm.props().hidden).toBeFalsy();
  });
});
