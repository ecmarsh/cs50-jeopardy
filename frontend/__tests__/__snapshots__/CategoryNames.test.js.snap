// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CategoryNames /> renders edit categories form and matches snapshot 1`] = `
<form
  className="noBg Form-qsq7s5-0 bLtDzK"
  data-test="category-names-form"
  onSubmit={[Function]}
>
  <h2>
    Category Names
  </h2>
  <DisplayError
    error={Object {}}
  />
  <fieldset
    aria-busy={false}
    disabled={false}
  >
    <div
      key="cat-1"
    >
      <h4>
        Single Jeopardy
      </h4>
      <label
        htmlFor="category-0"
        style={
          Object {
            "flexDirection": "row",
            "width": "100%",
          }
        }
      >
        <Utilities__TextMuted
          style={
            Object {
              "flexBasis": "25%",
              "textAlign": "center",
            }
          }
        >
          <StyledComponent
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "Utilities__TextMuted-s11g5i-2",
                  "isStatic": false,
                  "lastClassName": "biOppA",
                  "rules": Array [
                    "opacity:0.8;font-size:0.8em;display:inline;opacity:1;transition:opacity 0.3s ease-in-out;",
                    [Function],
                    " ",
                    [Function],
                    " ",
                    [Function],
                    " ",
                    [Function],
                  ],
                },
                "displayName": "Utilities__TextMuted",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "Utilities__TextMuted-s11g5i-2",
                "target": "span",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
            style={
              Object {
                "flexBasis": "25%",
                "textAlign": "center",
              }
            }
          >
            <span
              className="Utilities__TextMuted-s11g5i-2 bdrBqE"
              style={
                Object {
                  "flexBasis": "25%",
                  "textAlign": "center",
                }
              }
            >
              CATEGORY 
              1
              :
            </span>
          </StyledComponent>
        </Utilities__TextMuted>
        <input
          id="cat-1"
          name="category-0"
          onChange={[Function]}
          style={
            Object {
              "flexBasis": "60%",
              "width": "auto",
            }
          }
          type="text"
          value="omnis"
        />
      </label>
    </div>
    <div
      key="cat-2"
    >
      <label
        htmlFor="category-1"
        style={
          Object {
            "flexDirection": "row",
            "width": "100%",
          }
        }
      >
        <Utilities__TextMuted
          style={
            Object {
              "flexBasis": "25%",
              "textAlign": "center",
            }
          }
        >
          <StyledComponent
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "Utilities__TextMuted-s11g5i-2",
                  "isStatic": false,
                  "lastClassName": "biOppA",
                  "rules": Array [
                    "opacity:0.8;font-size:0.8em;display:inline;opacity:1;transition:opacity 0.3s ease-in-out;",
                    [Function],
                    " ",
                    [Function],
                    " ",
                    [Function],
                    " ",
                    [Function],
                  ],
                },
                "displayName": "Utilities__TextMuted",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "Utilities__TextMuted-s11g5i-2",
                "target": "span",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
            style={
              Object {
                "flexBasis": "25%",
                "textAlign": "center",
              }
            }
          >
            <span
              className="Utilities__TextMuted-s11g5i-2 bdrBqE"
              style={
                Object {
                  "flexBasis": "25%",
                  "textAlign": "center",
                }
              }
            >
              CATEGORY 
              2
              :
            </span>
          </StyledComponent>
        </Utilities__TextMuted>
        <input
          id="cat-2"
          name="category-1"
          onChange={[Function]}
          style={
            Object {
              "flexBasis": "60%",
              "width": "auto",
            }
          }
          type="text"
          value="nesciunt"
        />
      </label>
    </div>
    <div
      key="cat-3"
    >
      <label
        htmlFor="category-2"
        style={
          Object {
            "flexDirection": "row",
            "width": "100%",
          }
        }
      >
        <Utilities__TextMuted
          style={
            Object {
              "flexBasis": "25%",
              "textAlign": "center",
            }
          }
        >
          <StyledComponent
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "Utilities__TextMuted-s11g5i-2",
                  "isStatic": false,
                  "lastClassName": "biOppA",
                  "rules": Array [
                    "opacity:0.8;font-size:0.8em;display:inline;opacity:1;transition:opacity 0.3s ease-in-out;",
                    [Function],
                    " ",
                    [Function],
                    " ",
                    [Function],
                    " ",
                    [Function],
                  ],
                },
                "displayName": "Utilities__TextMuted",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "Utilities__TextMuted-s11g5i-2",
                "target": "span",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
            style={
              Object {
                "flexBasis": "25%",
                "textAlign": "center",
              }
            }
          >
            <span
              className="Utilities__TextMuted-s11g5i-2 bdrBqE"
              style={
                Object {
                  "flexBasis": "25%",
                  "textAlign": "center",
                }
              }
            >
              CATEGORY 
              3
              :
            </span>
          </StyledComponent>
        </Utilities__TextMuted>
        <input
          id="cat-3"
          name="category-2"
          onChange={[Function]}
          style={
            Object {
              "flexBasis": "60%",
              "width": "auto",
            }
          }
          type="text"
          value="occaecati"
        />
      </label>
    </div>
    <div
      key="cat-4"
    >
      <label
        htmlFor="category-3"
        style={
          Object {
            "flexDirection": "row",
            "width": "100%",
          }
        }
      >
        <Utilities__TextMuted
          style={
            Object {
              "flexBasis": "25%",
              "textAlign": "center",
            }
          }
        >
          <StyledComponent
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "Utilities__TextMuted-s11g5i-2",
                  "isStatic": false,
                  "lastClassName": "biOppA",
                  "rules": Array [
                    "opacity:0.8;font-size:0.8em;display:inline;opacity:1;transition:opacity 0.3s ease-in-out;",
                    [Function],
                    " ",
                    [Function],
                    " ",
                    [Function],
                    " ",
                    [Function],
                  ],
                },
                "displayName": "Utilities__TextMuted",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "Utilities__TextMuted-s11g5i-2",
                "target": "span",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
            style={
              Object {
                "flexBasis": "25%",
                "textAlign": "center",
              }
            }
          >
            <span
              className="Utilities__TextMuted-s11g5i-2 bdrBqE"
              style={
                Object {
                  "flexBasis": "25%",
                  "textAlign": "center",
                }
              }
            >
              CATEGORY 
              4
              :
            </span>
          </StyledComponent>
        </Utilities__TextMuted>
        <input
          id="cat-4"
          name="category-3"
          onChange={[Function]}
          style={
            Object {
              "flexBasis": "60%",
              "width": "auto",
            }
          }
          type="text"
          value="assumenda"
        />
      </label>
    </div>
    <div
      key="cat-5"
    >
      <label
        htmlFor="category-4"
        style={
          Object {
            "flexDirection": "row",
            "width": "100%",
          }
        }
      >
        <Utilities__TextMuted
          style={
            Object {
              "flexBasis": "25%",
              "textAlign": "center",
            }
          }
        >
          <StyledComponent
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "Utilities__TextMuted-s11g5i-2",
                  "isStatic": false,
                  "lastClassName": "biOppA",
                  "rules": Array [
                    "opacity:0.8;font-size:0.8em;display:inline;opacity:1;transition:opacity 0.3s ease-in-out;",
                    [Function],
                    " ",
                    [Function],
                    " ",
                    [Function],
                    " ",
                    [Function],
                  ],
                },
                "displayName": "Utilities__TextMuted",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "Utilities__TextMuted-s11g5i-2",
                "target": "span",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
            style={
              Object {
                "flexBasis": "25%",
                "textAlign": "center",
              }
            }
          >
            <span
              className="Utilities__TextMuted-s11g5i-2 bdrBqE"
              style={
                Object {
                  "flexBasis": "25%",
                  "textAlign": "center",
                }
              }
            >
              CATEGORY 
              5
              :
            </span>
          </StyledComponent>
        </Utilities__TextMuted>
        <input
          id="cat-5"
          name="category-4"
          onChange={[Function]}
          style={
            Object {
              "flexBasis": "60%",
              "width": "auto",
            }
          }
          type="text"
          value="porro"
        />
      </label>
    </div>
    <div
      key="cat-6"
    >
      <h4>
        Double Jeopardy
      </h4>
      <label
        htmlFor="category-5"
        style={
          Object {
            "flexDirection": "row",
            "width": "100%",
          }
        }
      >
        <Utilities__TextMuted
          style={
            Object {
              "flexBasis": "25%",
              "textAlign": "center",
            }
          }
        >
          <StyledComponent
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "Utilities__TextMuted-s11g5i-2",
                  "isStatic": false,
                  "lastClassName": "biOppA",
                  "rules": Array [
                    "opacity:0.8;font-size:0.8em;display:inline;opacity:1;transition:opacity 0.3s ease-in-out;",
                    [Function],
                    " ",
                    [Function],
                    " ",
                    [Function],
                    " ",
                    [Function],
                  ],
                },
                "displayName": "Utilities__TextMuted",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "Utilities__TextMuted-s11g5i-2",
                "target": "span",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
            style={
              Object {
                "flexBasis": "25%",
                "textAlign": "center",
              }
            }
          >
            <span
              className="Utilities__TextMuted-s11g5i-2 bdrBqE"
              style={
                Object {
                  "flexBasis": "25%",
                  "textAlign": "center",
                }
              }
            >
              CATEGORY 
              6
              :
            </span>
          </StyledComponent>
        </Utilities__TextMuted>
        <input
          id="cat-6"
          name="category-5"
          onChange={[Function]}
          style={
            Object {
              "flexBasis": "60%",
              "width": "auto",
            }
          }
          type="text"
          value="reprehenderit"
        />
      </label>
    </div>
    <div
      key="cat-7"
    >
      <label
        htmlFor="category-6"
        style={
          Object {
            "flexDirection": "row",
            "width": "100%",
          }
        }
      >
        <Utilities__TextMuted
          style={
            Object {
              "flexBasis": "25%",
              "textAlign": "center",
            }
          }
        >
          <StyledComponent
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "Utilities__TextMuted-s11g5i-2",
                  "isStatic": false,
                  "lastClassName": "biOppA",
                  "rules": Array [
                    "opacity:0.8;font-size:0.8em;display:inline;opacity:1;transition:opacity 0.3s ease-in-out;",
                    [Function],
                    " ",
                    [Function],
                    " ",
                    [Function],
                    " ",
                    [Function],
                  ],
                },
                "displayName": "Utilities__TextMuted",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "Utilities__TextMuted-s11g5i-2",
                "target": "span",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
            style={
              Object {
                "flexBasis": "25%",
                "textAlign": "center",
              }
            }
          >
            <span
              className="Utilities__TextMuted-s11g5i-2 bdrBqE"
              style={
                Object {
                  "flexBasis": "25%",
                  "textAlign": "center",
                }
              }
            >
              CATEGORY 
              7
              :
            </span>
          </StyledComponent>
        </Utilities__TextMuted>
        <input
          id="cat-7"
          name="category-6"
          onChange={[Function]}
          style={
            Object {
              "flexBasis": "60%",
              "width": "auto",
            }
          }
          type="text"
          value="iure"
        />
      </label>
    </div>
    <div
      key="cat-8"
    >
      <label
        htmlFor="category-7"
        style={
          Object {
            "flexDirection": "row",
            "width": "100%",
          }
        }
      >
        <Utilities__TextMuted
          style={
            Object {
              "flexBasis": "25%",
              "textAlign": "center",
            }
          }
        >
          <StyledComponent
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "Utilities__TextMuted-s11g5i-2",
                  "isStatic": false,
                  "lastClassName": "biOppA",
                  "rules": Array [
                    "opacity:0.8;font-size:0.8em;display:inline;opacity:1;transition:opacity 0.3s ease-in-out;",
                    [Function],
                    " ",
                    [Function],
                    " ",
                    [Function],
                    " ",
                    [Function],
                  ],
                },
                "displayName": "Utilities__TextMuted",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "Utilities__TextMuted-s11g5i-2",
                "target": "span",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
            style={
              Object {
                "flexBasis": "25%",
                "textAlign": "center",
              }
            }
          >
            <span
              className="Utilities__TextMuted-s11g5i-2 bdrBqE"
              style={
                Object {
                  "flexBasis": "25%",
                  "textAlign": "center",
                }
              }
            >
              CATEGORY 
              8
              :
            </span>
          </StyledComponent>
        </Utilities__TextMuted>
        <input
          id="cat-8"
          name="category-7"
          onChange={[Function]}
          style={
            Object {
              "flexBasis": "60%",
              "width": "auto",
            }
          }
          type="text"
          value="earum"
        />
      </label>
    </div>
    <div
      key="cat-9"
    >
      <label
        htmlFor="category-8"
        style={
          Object {
            "flexDirection": "row",
            "width": "100%",
          }
        }
      >
        <Utilities__TextMuted
          style={
            Object {
              "flexBasis": "25%",
              "textAlign": "center",
            }
          }
        >
          <StyledComponent
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "Utilities__TextMuted-s11g5i-2",
                  "isStatic": false,
                  "lastClassName": "biOppA",
                  "rules": Array [
                    "opacity:0.8;font-size:0.8em;display:inline;opacity:1;transition:opacity 0.3s ease-in-out;",
                    [Function],
                    " ",
                    [Function],
                    " ",
                    [Function],
                    " ",
                    [Function],
                  ],
                },
                "displayName": "Utilities__TextMuted",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "Utilities__TextMuted-s11g5i-2",
                "target": "span",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
            style={
              Object {
                "flexBasis": "25%",
                "textAlign": "center",
              }
            }
          >
            <span
              className="Utilities__TextMuted-s11g5i-2 bdrBqE"
              style={
                Object {
                  "flexBasis": "25%",
                  "textAlign": "center",
                }
              }
            >
              CATEGORY 
              9
              :
            </span>
          </StyledComponent>
        </Utilities__TextMuted>
        <input
          id="cat-9"
          name="category-8"
          onChange={[Function]}
          style={
            Object {
              "flexBasis": "60%",
              "width": "auto",
            }
          }
          type="text"
          value="dolorem"
        />
      </label>
    </div>
    <div
      key="cat-10"
    >
      <label
        htmlFor="category-9"
        style={
          Object {
            "flexDirection": "row",
            "width": "100%",
          }
        }
      >
        <Utilities__TextMuted
          style={
            Object {
              "flexBasis": "25%",
              "textAlign": "center",
            }
          }
        >
          <StyledComponent
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "Utilities__TextMuted-s11g5i-2",
                  "isStatic": false,
                  "lastClassName": "biOppA",
                  "rules": Array [
                    "opacity:0.8;font-size:0.8em;display:inline;opacity:1;transition:opacity 0.3s ease-in-out;",
                    [Function],
                    " ",
                    [Function],
                    " ",
                    [Function],
                    " ",
                    [Function],
                  ],
                },
                "displayName": "Utilities__TextMuted",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "Utilities__TextMuted-s11g5i-2",
                "target": "span",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
            style={
              Object {
                "flexBasis": "25%",
                "textAlign": "center",
              }
            }
          >
            <span
              className="Utilities__TextMuted-s11g5i-2 bdrBqE"
              style={
                Object {
                  "flexBasis": "25%",
                  "textAlign": "center",
                }
              }
            >
              CATEGORY 
              10
              :
            </span>
          </StyledComponent>
        </Utilities__TextMuted>
        <input
          id="cat-10"
          name="category-9"
          onChange={[Function]}
          style={
            Object {
              "flexBasis": "60%",
              "width": "auto",
            }
          }
          type="text"
          value="atque"
        />
      </label>
    </div>
    <div
      key="cat-11"
    >
      <h4>
        Final Jeopardy
      </h4>
      <label
        htmlFor="category-10"
        style={
          Object {
            "flexDirection": "row",
            "width": "100%",
          }
        }
      >
        <Utilities__TextMuted
          style={
            Object {
              "flexBasis": "25%",
              "textAlign": "center",
            }
          }
        >
          <StyledComponent
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "Utilities__TextMuted-s11g5i-2",
                  "isStatic": false,
                  "lastClassName": "biOppA",
                  "rules": Array [
                    "opacity:0.8;font-size:0.8em;display:inline;opacity:1;transition:opacity 0.3s ease-in-out;",
                    [Function],
                    " ",
                    [Function],
                    " ",
                    [Function],
                    " ",
                    [Function],
                  ],
                },
                "displayName": "Utilities__TextMuted",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "Utilities__TextMuted-s11g5i-2",
                "target": "span",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
            style={
              Object {
                "flexBasis": "25%",
                "textAlign": "center",
              }
            }
          >
            <span
              className="Utilities__TextMuted-s11g5i-2 bdrBqE"
              style={
                Object {
                  "flexBasis": "25%",
                  "textAlign": "center",
                }
              }
            >
              CATEGORY 
              11
              :
            </span>
          </StyledComponent>
        </Utilities__TextMuted>
        <input
          id="cat-11"
          name="category-10"
          onChange={[Function]}
          style={
            Object {
              "flexBasis": "60%",
              "width": "auto",
            }
          }
          type="text"
          value="ea"
        />
      </label>
    </div>
    <div
      style={
        Object {
          "margin": "0 auto",
        }
      }
    >
      <Utilities__TextMuted
        data-test="category-names-save-confirmation"
        hidden={true}
      >
        <StyledComponent
          data-test="category-names-save-confirmation"
          forwardedComponent={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "attrs": Array [],
              "componentStyle": ComponentStyle {
                "componentId": "Utilities__TextMuted-s11g5i-2",
                "isStatic": false,
                "lastClassName": "biOppA",
                "rules": Array [
                  "opacity:0.8;font-size:0.8em;display:inline;opacity:1;transition:opacity 0.3s ease-in-out;",
                  [Function],
                  " ",
                  [Function],
                  " ",
                  [Function],
                  " ",
                  [Function],
                ],
              },
              "displayName": "Utilities__TextMuted",
              "foldedComponentIds": Array [],
              "render": [Function],
              "styledComponentId": "Utilities__TextMuted-s11g5i-2",
              "target": "span",
              "toString": [Function],
              "warnTooManyClasses": [Function],
              "withComponent": [Function],
            }
          }
          forwardedRef={null}
          hidden={true}
        >
          <span
            className="Utilities__TextMuted-s11g5i-2 biOppA"
            data-test="category-names-save-confirmation"
            hidden={true}
          >
            Category Names Saved!
          </span>
        </StyledComponent>
      </Utilities__TextMuted>
      <Button
        primary={true}
        style={
          Object {
            "margin": "1rem auto 2rem",
          }
        }
        type="submit"
      >
        <StyledComponent
          forwardedComponent={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "attrs": Array [],
              "componentStyle": ComponentStyle {
                "componentId": "Button-sc-1f1w3gq-0",
                "isStatic": false,
                "lastClassName": "jSaPjy",
                "rules": Array [
                  "font-family:",
                  [Function],
                  ";color:white;border-radius:0.5rem;text-transform:uppercase;letter-spacing:1px;transition:.2s ease;&:hover{cursor:pointer;}&:focus{cursor:pointer;outline:",
                  [Function],
                  " auto 3px;}",
                  [Function],
                  " ",
                  [Function],
                  " ",
                  [Function],
                  " ",
                  [Function],
                  " ",
                  [Function],
                  " &.button--sound{font-weight:bolder;text-decoration:line-through;&::after{content:\\"❌\\";padding-left:3px;}",
                  [Function],
                  "}",
                  [Function],
                ],
              },
              "displayName": "Button",
              "foldedComponentIds": Array [],
              "render": [Function],
              "styledComponentId": "Button-sc-1f1w3gq-0",
              "target": "button",
              "toString": [Function],
              "warnTooManyClasses": [Function],
              "withComponent": [Function],
            }
          }
          forwardedRef={null}
          primary={true}
          style={
            Object {
              "margin": "1rem auto 2rem",
            }
          }
          type="submit"
        >
          <button
            className="Button-sc-1f1w3gq-0 jSaPjy"
            style={
              Object {
                "margin": "1rem auto 2rem",
              }
            }
            type="submit"
          >
            Save Category Names
          </button>
        </StyledComponent>
      </Button>
      <Utilities__Divider>
        <StyledComponent
          forwardedComponent={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "attrs": Array [],
              "componentStyle": ComponentStyle {
                "componentId": "Utilities__Divider-s11g5i-5",
                "isStatic": true,
                "lastClassName": "jrubw",
                "rules": Array [
                  "width:80%;opacity:0.5;margin:1rem auto;",
                ],
              },
              "displayName": "Utilities__Divider",
              "foldedComponentIds": Array [],
              "render": [Function],
              "styledComponentId": "Utilities__Divider-s11g5i-5",
              "target": "hr",
              "toString": [Function],
              "warnTooManyClasses": [Function],
              "withComponent": [Function],
            }
          }
          forwardedRef={null}
        >
          <hr
            className="Utilities__Divider-s11g5i-5 jrubw"
          />
        </StyledComponent>
      </Utilities__Divider>
    </div>
  </fieldset>
</form>
`;
